default/python/AI/AIstate.py

```diff
@@ -62,18 +62,10 @@ def convert_to_version(state, version):
     if version != current_version + 1:
         raise ConversionError("Can't skip a compatibility version when converting AI savegame.")
 
-    if version == 0:
-        pass  # only version number added
-    elif version == 1:
-        try:
-            state['_aggression'] = state['character'].get_trait(Aggression).key
-        except Exception as e:
-            raise ConversionError("Error when converting to compatibility version 1: "
-                                  "Can't find aggression in character module. Exception thrown was: " + e.message)
-    elif version == 2:
-        # some dicts which used only the keys were changed to sets
-        for var_name in ['visInteriorSystemIDs', 'exploredSystemIDs', 'visBorderSystemIDs', 'unexploredSystemIDs']:
-            state[var_name] = set(state.get(var_name, {}).keys())
+    # Starting with version 3, we switched from pickle to json-style encoding
+    # Do not try to load an older savegame even if it magically passed the encoder.
+    if version <= 3:
+        raise ConversionError("The AI savegame version is no longer supported.")
```

Comment:
It is not clear for me at this point if we will just cleanup AI state (Is message like "AI state will lost memory and continue game" show to user) or stop game (in that case it will be useful to add game commit/ version that should be rollback to finish game.)

Reply:
Do both...? Let the player continue with the current program version if they want, but add a note that the loaded save file was created with version X so that the player can revert if they want.

Reply:
The game will automatically continue, simply recreating a new AIstate rather than using the old stale-versioned (and potentially unsafe) one, and I think there is no need (and no ready way) to pause the game loading and query the user about what they want to do about it.   (As Morlic hinted with the 'magically passed' reference, I don't see how any such older state could even pass through the new encoder, and if it somehow did it would be hopelessly garbled).

The comment line could be somewhat clearer though, rather than referring to no loading an "older savegame" I think it would be more clear to refer to an "older version aistate", and the ConversionError message, which will be presented to the player in the chat window, could more clearly explain that the AIstate will be recreated fresh.